.github/workflows/main.yml

```diff
@@ -1,53 +1,80 @@
 on: [push, pull_request]
+env:
+  PYTHONWARNINGS: "ignore"
+  PIP_ROOT_USER_ACTION: "ignore"   # For local testing using act-cli
+  PIP_NO_WARN_SCRIPT_LOCATION: "0" # For local testing using act-cli
 
 jobs:
   test:
     strategy:
+      # max-parallel: 1  # See: https://github.com/xenserver/python-libs/pull/26#discussion_r1179482169
       matrix:
         include:
+        - python-version: '3.11'
+          os: ubuntu-latest
+        - python-version: '3.10'
+          os: ubuntu-latest
         - python-version: '2.7'
           os: ubuntu-20.04
-        - python-version: '3'
+        - python-version: '3.9'
+          os: ubuntu-latest
+        - python-version: '3.8'
           os: ubuntu-latest
+        - python-version: '3.7'
+          os: ubuntu-latest
+        - python-version: '3.6'
+          os: ubuntu-20.04
     runs-on: ${{ matrix.os }}
 
     steps:
       - uses: actions/checkout@v3
         with:
-          fetch-depth: 0
+          fetch-depth: 0  # Needed by diff-cover to get the changed lines: origin/master..HEAD
       - name: Set up Python ${{ matrix.python-version }}
         uses: actions/setup-python@v4
         with:
           python-version: ${{ matrix.python-version }}
 
-      - name: Install dependencies
+      - name: Provide Pylint Summary and Report on the GitHub Action's info page
+        if: ${{ matrix.python-version == '3.10' }}
         run: |
-          python -m pip install --upgrade pip
-          pip install -r requirements-dev.txt
-          pip install -e .
+          pip install -q pylint pandas tabulate 2>&1|grep -v root ||: # Hide warning for root in local container
+          tree=$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/tree/$GITHUB_REF_NAME
+          [ -n "$GITHUB_STEP_SUMMARY" ] || GITHUB_STEP_SUMMARY=pylint-summary-table.md
+          ./gen_gh_pylint_report.py xcp $GITHUB_STEP_SUMMARY $tree
 
       - name: Test
         run: |
-          pytest --cov -rP
-          coverage xml
-          coverage html -d htmlcov-${{ matrix.python-version }}
-          coverage html -d htmlcov-tests-${{ matrix.python-version }} --include="tests/*"
-          diff-cover --compare-branch=origin/master --html-report coverage-diff-${{ matrix.python-version }}.html coverage.xml
+          rm -f coverage.xml
+          pip -q install tox tox-gh-actions
+          tox --recreate
 
-      - name: Pylint
-        run: |
-          pylint --version
-          pylint --exit-zero xcp/ tests/ setup.py
-          pylint --exit-zero --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" xcp/ tests/ setup.py > pylint.txt
-          diff-quality --compare-branch=origin/master --violations=pylint --html-report pylint-diff-${{ matrix.python-version }}.html pylint.txt
+      - name: Upload coverage reports to Codecov
+        if: ${{ matrix.python-version == '3.11' }}
+        uses: codecov/codecov-action@v3
 
       - uses: actions/upload-artifact@v3
```

Comment:
Why is this step needed? coverage is already reported to codecov in the previous step

Reply:
Thanks, I'm happy to remove this now redundant step.